### ISSUE TITLE
Isolates api throws error

### ISSUE THREAD

**Issue by VadimOsovsky** [simolus3/moor#242]
```
E/flutter (21654): Invalid argument(s): Illegal argument in isolate message : (object is a StackTrace)
E/flutter (21654): #0      _SendPortImpl._sendInternal (dart:isolate-patch/isolate_patch.dart:225:70)
E/flutter (21654): #1      _SendPortImpl.send (dart:isolate-patch/isolate_patch.dart:209:5)
E/flutter (21654): #2      IsolateCommunication._send (package:moor/src/runtime/isolate/communication.dart:118:14)
E/flutter (21654): #3      IsolateCommunication.respondError (package:moor/src/runtime/isolate/communication.dart:128:5)
E/flutter (21654): #4      IsolateCommunication.setRequestHandler.<anonymous closure> (package:moor/src/runtime/isolate/communication.dart:146:9)
E/flutter (21654): #5      _RootZone.runUnaryGuarded (dart:async/zone.dart:1316:10)
E/flutter (21654): #6      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:338:11)
E/flutter (21654): #7      _DelayedData.perform (dart:async/stream_impl.dart:593:14)
E/flutter (21654): #8      _StreamImplEvents.handleNext (dart:async/stream_impl.dart:709:11)
E/flutter (21654): #9      _PendingEvents.schedule.<anonymous closure> (dart:async/stream_impl.dart:669:7)
E/flutter (21654): #10     _microtaskLoop (dart:async/schedule_microtask.dart:43:21)
E/flutter (21654): #11     _startMicrotaskLoop (dart:async/schedule_microtask.dart:52:5)
E/flutter (21654): #12     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:118:13)
E/flutter (21654): #13     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:175:5)
```

**Comment by simolus3** [comment]
Thanks for the report. The channel should actually serialize and deserialize stack traces. This could be caused by an exception which has a reference to a `StackTrace` itself.

Can you manually set a breakpoint on `package:moor/src/runtime/isolate/communication.dart:146` (the catch clause). You can then see the exception and causing trace. Posting them would help to identify the underlying issue.

**Comment by felixjunghans** [comment]
I got the same Error, when I try to run transactions inside an isolate. I set a breakpoint and got the following Error.

#0      Object.noSuchMethod (dart:core-patch/object_patch.dart:53:5)
#1      _MoorServer._transactionControl (package:moor/src/runtime/isolate/server.dart:99:28)
#2      _MoorServer._handleRequest (package:moor/src/runtime/isolate/server.dart:63:14)
#3      IsolateCommunication.setRequestHandler.<anonymous closure> (package:moor/src/runtime/isolate/communication.dart:146:31)
#4      _RootZone.runUnaryGuarded (dart:async/zone.dart:1316:10)
#5      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:338:11)
#6      _DelayedData.perform (dart:async/stream_impl.dart:593:14)
#7      _StreamImplEvents.handleNext (dart:async/stream_impl.dart:709:11)
#8      _PendingEvents.schedule.<anonymous closure> (dart:async/stream_impl.dart:669:7)
#9      _microtaskLoop (dart:async/schedule_microtask.dart:43:21)
#10     _startMicrotaskLoop (dart:async/schedule_microtask.dart:52:5)
#11     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:118:13)
#12     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:175:5)

**Comment by felixjunghans** [comment]
I think I could find the bug. But since I'm using the package only for a short time and I'm not sure if I understand everything correctly, here is my solution proposal.

@simolus3 Maybe you can see if this is the solution.

I noticed that _transactionId is always null. This results in the fact that no transaction can be found and transaction.send leads to the error posted above. 

I then changed the transaction method as follows. This solution solved the problem for me.

`  Future<T> transaction<T>(Future<T> Function() action) async {
    final resolved = _resolvedEngine;
    if (resolved is Transaction) {
      return action();
    }

    final executor = resolved.executor;
    return await executor.doWhenOpened((executor) async {
      final transactionExecutor = executor.beginTransaction();

      
      **await transactionExecutor.ensureOpen();**
      final transaction = Transaction(this, transactionExecutor);

      return _runEngineZoned(transaction, () async {
        var success = false;
        try {
          final result = await action();
          success = true;
          return result;
        } catch (e) {
          await transactionExecutor.rollback();

          // pass the exception on to the one who called transaction()
          rethrow;
        } finally {
          if (success) {
            // complete() will also take care of committing the transaction
            await transaction.complete();
          }
        }
      });
    });
  }`

**Comment by simolus3** [comment]
Good analysis @felixjunghans! The `_transactionId` not being set is a bug indeed. There were some other bugs with transactions on isolates not working properly, I fixed those with c2ec06c1de063a3a1c1a05476e6a31c773ed0938. This hasn't been released yet, I'll try to release the fix as part of moor 2.3 next week.

**Comment by felixjunghans** [comment]
Nice! Thanks for your reply and your work on this package.

**Comment by simolus3** [comment]
The problem with `_transactionId` should be fixed in moor 2.3, which I just released. If you run into any other problems, just let me know!
